framework/src/media/audio/audio_manager.c

```diff
@@ -500,12 +546,44 @@ int start_audio_stream_in(void *data, uint32_t frames)
 		}
 
 		g_audio_in_cards[g_actual_audio_in_card_id].status = AUDIO_CARD_RUNNING;
-		audvdbg("Resume the input audio card!!\n");
+		auddbg("Resume the input audio card!!\n");
 	}
 
 	do {
-		ret = pcm_readi(pcm, data, frames);
-		audvdbg("Read %d frames\n", ret);
+		if(cur_card->resample.necessary) {
+			unsigned int frames_to_read = (unsigned int)((float)frames * cur_card->resample.ratio);
+			if(frames_to_read > (get_input_frame_count() * cur_card->resample.ratio)) {
+				frames_to_read = (get_input_frame_count() * cur_card->resample.ratio);
+			}
+
+			auddbg("Read frame for resample: %u to %u\n", frames_to_read, frames);
+
+			ret = pcm_readi(pcm, cur_card->resample.readbuffer, frames_to_read);
+
+			src_data_t srcData;
+            srcData.data_in = cur_card->resample.readbuffer;
+            srcData.input_frames = ret;
+            srcData.channels_num = 2; // Todo: Consider Mono/Stereo
+            srcData.origin_sample_rate = cur_card->resample.from;
+			srcData.origin_sample_width = SAMPLE_WIDTH_16BITS;
+            srcData.desired_sample_rate = cur_card->resample.to;
+			srcData.desired_sample_width = SAMPLE_WIDTH_16BITS;
+
+			if(src_simple(cur_card->resample.handle, &srcData) != SRC_ERR_NO_ERROR) {
+				auddbg("Fail to resample to %u from %u\n", srcData.desired_sample_rate, srcData.origin_sample_rate);
+				return AUDIO_MANAGER_RESAMPLE_FAIL;
+			}
+
+			auddbg("Record resampled frames: from %d to %d\n", ret, srcData.output_frames_gen);
+			ret = srcData.output_frames_gen;
+
+			memcpy(data, srcData.data_out, get_input_frames_byte_size(srcData.output_frames_gen));
```

Comment:
if get frames contains channel currently, below is correct
memcpy(data, srcData.data_out, srcData.output_frames_gen & SAMPLE_WIDTH_16BITS);

Reply:
Resampler is changed. output_frames_gen doesn't contain channel count anymore.

Reply:
ok it's good for us.